### ISSUE TITLE
RFC: Extension proposal

### ISSUE THREAD

**Issue by TillaTheHun0** [hyper63/hyper-ext-counter#1]
Proposal for a "poor man's" extension system

The idea is to create higher order functions that take a hyper instance, then return a hyper instance with the extension attached. Then the extensions can be composed together.

`const hyper = compose(counter, addAndIndex, logger)(connect(process.env.HYPER))`

**Comment by twilson63** [comment]
The point is to create a way to experiment on composable interfaces without committing to a specific plugin system out of the gate and possibly get it wrong. We could potentially leverage npm download counts to discover usability, if extensions are becoming popular then we consider adding them into `hyper-connect` 

Another way we can approach it is like `pouchdb`

```js
const hyper = connect(process.env.HYPER)
hyper.plugin(counter)
hyper.plugin(addAndIndex)
hyper.plugin(logger)
```

Then the plugin function would simply extend the hyper object, my concerns with this approach are the implicit mutation of the hyper object and typescript IntelliSense. 

I prefer to embrace functional semantics if possible.


For starters

**Comment by TillaTheHun0** [comment]
I like the HoF approach and the ability to compose a set of extensions 👍 . Seems like it could be non obtrusive way for extending `connect`, with low barrier to entry; it's simple to write one of these extensions, no new api, just `connect`.

~This seems separate from the plugin engine ideas we've discussed. This seems more like `mixins` for hyper-connect, as each extension adds function(s) onto `ext`. I suppose an extension could call other extensions, but there is no orchestrator involved (ie. engine) is the point. Do I have that right?~ **Answer**: yes this is precedes plugins and is meant to identify patterns that could potentially inform a plugin engine design.

I could see extensions colliding on the `ext` namespace. Perhaps each extension _must_ add all of it's entrypoints under a barrel on `ext`, basically what you've done with `ext.counter`. Don't know if that was the idea for it or not. Should we do something if there _is_ a collision? a warning? an error? nothing?

**Comment by TillaTheHun0** [comment]
> Then the plugin function would simply extend the hyper object, my concerns with this approach are the implicit mutation of the hyper object and typescript IntelliSense.

I agree with this point. To my point about collisions above, perhaps we offer a `compose` that detects collisions? If we want to do that sort of thing.

**Comment by TillaTheHun0** [comment]
Could extensions overwrite one or multiple `connect` methods? Or do they _have_ to modify `ext`? Should we `freeze` the `connect` "core" methods to prevent this?

**Comment by twilson63** [comment]
> > Then the plugin function would simply extend the hyper object, my concerns with this approach are the implicit mutation of the hyper object and typescript IntelliSense.
> 
> I agree with this point. To my point about collisions above, perhaps we offer a `compose` that detects collisions? If we want to do that sort of thing.

Yes, that is a good idea

```
import { compose, connect } from 'hyper-connect'
const hyper = compose(a,b,c)(connect(process.env.HYPER))
// { ..., ext: ... }
```

Then we automatically compose each fn over the ext - using lenses


And report errors for conflicts, collisions etc
